apis/v1alpha2/validation/gateway.go

```diff
@@ -44,7 +58,25 @@ func validateGatewaySpec(spec *gatewayv1a2.GatewaySpec, path *field.Path) field.
 // to the Gateway API specification.
 func validateGatewayListeners(listeners []gatewayv1a2.Listener, path *field.Path) field.ErrorList {
 	// TODO [danehans]: Add additional validation of listener fields.
-	return validateListenerHostname(listeners, path)
+	var errs field.ErrorList
+	errs = append(errs, validateListenerTLSConfig(listeners, path)...)
+	errs = append(errs, validateListenerHostname(listeners, path)...)
+	return errs
+}
+
+func validateListenerTLSConfig(listeners []gatewayv1a2.Listener, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	for i, l := range listeners {
+		if isProtocolInSubset(l.Protocol, nonTlsProtocols) && l.TLS != nil {
+			errs = append(errs, field.Forbidden(path.Index(i).Child("tls"), fmt.Sprintf("should be empty for protocol %v", l.Protocol)))
+		}
+	}
+	return errs
+}
+
+func isProtocolInSubset(protocol gatewayv1a2.ProtocolType, set map[gatewayv1a2.ProtocolType]struct{}) bool {
```

Comment:
can we use the standard `set` package? the code will be more understandable.

Reply:
Unfortunately the k8s set package only works with a few primitive types (including string). In some other cases like EndpointSlice controller [we've created sets based on that interface that support specific custom types](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/util/endpointslice/endpointset.go). We could do that here, but it may be too complex for this specific use case.